Sources/GRPCCore/GRPCClient.swift

```diff
@@ -0,0 +1,823 @@
+    /// The client is running and can send RPCs. Can transition to `stopping`.
+    case running
+    /// The client is stopping and no new RPCs will be sent. Existing RPCs may run to
+    /// completion. May transition to `stopped`.
+    case stopping
+    /// The client has stopped, no RPCs are in flight and no more will be accepted. This state
+    /// is terminal.
+    case stopped
+  }
+
+  /// Underlying storage for the client.
+  private struct Storage {
+    var interceptors: Interceptors
+    var methodConfigurationOverrides: MethodConfigurationRegistry
+    var state: State
+
+    init() {
+      self.interceptors = Interceptors()
+      self.methodConfigurationOverrides = MethodConfigurationRegistry()
+      self.state = .notStarted
+    }
+  }
+
+  private let storage: LockedValueBox<Storage>
+
+  /// The transport which provides a bidirectional communication channel with the server.
+  private let transport: ClientTransport
+
+  /// Creates a new client with no resources.
+  ///
+  /// You can add resources to the client via ``interceptors-swift.property`` and
+  /// ``methodConfigurationOverrides-swift.property``, and start the client by calling ``run()``.
+  ///
+  /// - Note: Any changes to resources after ``run()`` has been called will be ignored.
+  ///
+  /// - Parameter transport: The ``ClientTransport`` to be used for this ``GRPCClient``.
+  public init(transport: ClientTransport) {
+    self.storage = LockedValueBox(Storage())
+    self.transport = transport
+  }
+
+  /// Start the client.
+  ///
+  /// This is a long-running task that will return once ``close()`` has been called and all in-flight RPCs
+  /// finished executing.
+  ///
+  /// If you need to immediately stop all work, cancel the task executing this method.
+  public func run() async throws {
+    try self.storage.withLockedValue { storage in
+      switch storage.state {
+      case .notStarted:
+        storage.state = .running
+      case .running:
+        throw ClientError(
+          code: .clientIsAlreadyRunning,
+          message: "The client is already running and can only be started once."
+        )
+      case .stopping, .stopped:
+        throw ClientError(
+          code: .clientIsStopped,
+          message: "The client has stopped and can only be started once."
+        )
+      }
+    }
+
+    // When we exit this function we must have stopped.
+    defer {
+      self.storage.withLockedValue { $0.state = .stopped }
+    }
+
+    try await withThrowingTaskGroup(of: Void.self) { group in
+      group.addTask {
+        try await self.transport.connect(lazily: false)
+      }
+      try await group.next()
+    }
+  }
+
+  /// Close the client.
+  ///
+  /// The transport will be closed: this means that it will be given enough time to wait for in-flight RPCs to
+  /// finish executing, but no new RPCs will be accepted.
+  /// You can cancel the task executing ``run()`` if you want to immediately stop all work.
+  public func close() {
```

Comment:
IMO `close` is the wrong name here and we should follow the lifecycle naming of `triggerGracefulShutdown`

Reply:
Why do you think it's the wrong name?

Reply:
> Why do you think it's the wrong name?

At least coming from a NIO world or an FD world `close` means "close right now". However, from the documentation here it sounds way more like a graceful shutdown than an immediate close. Furthermore, we are kinda trying to establish the pattern of `gracefulShutdown` with `ServiceLifecycle` so I think aligning on a common name is beneficial for the ecosystem.